flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSource.java

```diff
@@ -208,29 +252,40 @@ public void close() throws Exception {
 		}
 	}
 
+	/**
+	 * Parse and collects the body of the an AMQP message.
+	 *
+	 * <p>If any of the constructors with the {@link DeserializationSchema} class was used to construct the source
+	 * it uses the {@link DeserializationSchema#deserialize(byte[])} to parse the body of the AMQP message.
+	 *
+	 * <p>If any of the constructors with the {@link RMQDeserializationSchema } class was used to construct the source it uses the
+	 * {@link RMQDeserializationSchema#processMessage(Envelope, AMQP.BasicProperties, byte[], RMQDeserializationSchema.RMQCollector collector)}
+	 * method of that provided instance.
+	 *
+	 * @param delivery the AMQP {@link QueueingConsumer.Delivery}
+	 * @param collector a {@link RMQCollectorImpl} to collect the data
+	 * @throws IOException
+	 */
+	protected void processMessage(QueueingConsumer.Delivery delivery, RMQDeserializationSchema.RMQCollector collector) throws IOException {
+		AMQP.BasicProperties properties = delivery.getProperties();
+		byte[] body = delivery.getBody();
+
+		if (deliveryDeserializer != null){
+			Envelope envelope = delivery.getEnvelope();
+			deliveryDeserializer.processMessage(envelope, properties, body, collector);
+		} else {
+			collector.setMessageIdentifiers(properties.getCorrelationId(), delivery.getEnvelope().getDeliveryTag());
+			collector.collect(schema.deserialize(body));
+		}
+	}
```

Comment:
If we wrap the schema and go with the `reset` method, this could be simplified to:

```java
protected void processMessage(QueueingConsumer.Delivery delivery, RMQDeserializationSchema.RMQCollector<OUT> collector) throws IOException {
		AMQP.BasicProperties properties = delivery.getProperties();
		byte[] body = delivery.getBody();
		Envelope envelope = delivery.getEnvelope();
        // ensure the collector knows a new message is being processed
		collector.reset();
		schema.deserialize(envelope, properties, body, collector);
	}
```

Reply:
actually after your awesome suggestion of wrapping the deserilaztion schema into the new interface this method is now:

```
		AMQP.BasicProperties properties = delivery.getProperties();
		byte[] body = delivery.getBody();
		Envelope envelope = delivery.getEnvelope();

		deliveryDeserializer.deserialize(envelope, properties, body, collector);
```
Hence my comment on top "Elegant Solution !"